keps/sig-auth/0000-20190116-service-account-key-pipereader.md

```diff
@@ -0,0 +1,129 @@
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+The Kubernetes API server has always read service account keys from disk as the process starts. As the API server can now verify and issue projected volume tokens, it would be advantageous to read these signing and verifying keys from a unix socket rather than a file.
+
+## Motivation
+
+For operators who want to regularly rotate the signing and verifying key files for projected volume tokens, the API server must be restarted in order to use a new key. To facilitate easy key rotation, this KEP includes an proposal for a grpc API to read keys.
+
+### Goals
+
+- A new API for reading signing and verifiying keys over a unix socket
+- Preserve existing behavior and performance for keys not read over a socket
+
+### Non-Goals
+
+- Reading TLS serving certificates and key from a socket or reloading of the API server with new cert and key
+- Reading any other certificates from a file
+- Supporting out-of-process JWT signing and verifying
+
+## Proposal
+
+### Preserve existing behavior
+
+The API server flags `--service-account-key-file` and `--service-account-signing-key-file` will continue be used for reading from files. The API server will detect if the provided files are regular files to be read once at process startup, or unix sockets that should be read on each token request using a new API.
+
+### Updates to API server token generation
+
+As of Kubernetes v1.13.2, the API server uses the functions `JWTTokenGenerator` and `JWTTokenAuthenticator`. New types that implement the `TokenGenerator` interface and support token validation will be added to `k8s.io/kubernetes/pkg/serviceaccount/`.
+
+### New API
+
+I'm proposing creating a new versioned grpc API under `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/util/flag/keyreader`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`
+
+```proto
+syntax = "proto3";
+
+package v1beta1;
+
+// This service defines the public APIs for remote key provider.
+service JWTKeyService {
+  // GetSigningKey returns a private key and a key id
+  rpc GetSigningKey(SigningKeyRequest) returns (SigningKeyReply) {}
+  // GetVerifyingKeys returns a list of public keys
+  rpc GetVerifyingKeys(VerifyingKeysRequest) returns (VerifyingKeysReply) {}
+}
+
+message SigningKeyRequest {
+}
+
+message SigningKeyReply {
+  // The PEM encoded bytes containing a private key
+  bytes private_key = 1;
+  // The key ID to sign tokens with
+  string key_id= 2;
+}
+
+message VerifyingKeysRequest {
+}
+
+message PublicKey {
+  // A public key
+  bytes public_key = 1;
+  // The key ID to verify tokens
+  string key_id = 2;
+}
+
+message VerifyingKeysReply {
+  // A list of public signing keys
+  repeated PublicKey public_keys = 1;
```

Comment:
Does it make sense to differentiate the active key?

Reply:
I've since broken this out into 2 calls, `GetPublicKey()` and `ListPublicKeys()`